easybuild/easyblocks/p/pytorch.py

```diff
@@ -0,0 +1,230 @@
+        self.testinstall = True
+        self.tmpdir = tempfile.mkdtemp(suffix='-pytorch-build')
+
+    def fetch_step(self, skip_checksums=False):
+        super(EB_PyTorch, self).fetch_step(skip_checksums)
+        # Resolve tests early to avoid failures later. Use obtain_file if path is not absolute
+        tests = [test if os.path.isabs(test) else self.obtain_file(test) for test in self.cfg['tests']]
+        self.cfg['tests'] = tests
+
+    @staticmethod
+    def get_dependency_options_for_version(pytorch_version):
+        """
+        PyTorch can enable some functionality based on available software or use system software instead of a submodule
+        This returns EasyBuild names of that and the flag that should be used when the dependency is found
+
+        The result is a list of tuples (enable_flag, eb_name)
+        """
+        pytorch_version = LooseVersion(pytorch_version)
+
+        def is_version_ok(version_range):
+            """Return True if the PyTorch version to be installed matches the version_range"""
+            min_version, max_version = version_range.split(':')
+            result = True
+            if min_version and pytorch_version < LooseVersion(min_version):
+                result = False
+            if max_version and pytorch_version >= LooseVersion(max_version):
+                result = False
+            return result
+
+        available_libs = (
+            # Format: (PyTorch flag to enable, EB name, '<min version>:<exclusive max version>')
+            # Use `None` for the EB name if no known EC exists
+            ('USE_FFMPEG=1', 'FFmpeg', '1.0.0:'),
+            ('USE_GFLAGS=1', 'gflags', '1.0.0:'),
+            ('USE_GLOG=1', 'glog', '1.0.0:'),
+
+            # For system libs check CMakeLists.txt, below `if(USE_SYSTEM_LIBS)`, order kept here
+            # NCCL handled specially as other env variables are requires for it
+            ('USE_SYSTEM_CPUINFO=1', None, '1.6.0:'),
+            ('USE_SYSTEM_SLEEF=1', None, '1.6.0:'),
+            ('USE_SYSTEM_GLOO=1', None, '1.6.0:'),
+            ('BUILD_CUSTOM_PROTOBUF=0', 'protobuf', '1.2.0:'),
+            ('USE_SYSTEM_EIGEN_INSTALL=1', 'Eigen', '1.0.0:'),
+            ('USE_SYSTEM_FP16=1', None, '1.6.0:'),
+            ('USE_SYSTEM_PTHREADPOOL=1', None, '1.6.0:'),
+            ('USE_SYSTEM_PSIMD=1', None, '1.6.0:'),
+            ('USE_SYSTEM_FXDIV=1', None, '1.6.0:'),
+            ('USE_SYSTEM_BENCHMARK=1', None, '1.6.0:'),  # Google Benchmark
+            ('USE_SYSTEM_ONNX=1', None, '1.6.0:'),
+            ('USE_SYSTEM_XNNPACK=1', None, '1.6.0:'),
+        )
+        return [(enable_opt, dep_name) for enable_opt, dep_name, version_range in available_libs
+                if is_version_ok(version_range)]
+
+    def configure_step(self):
+        """Custom configure procedure for PyTorch."""
+        super(EB_PyTorch, self).configure_step()
+
+        options = ['PYTORCH_BUILD_VERSION=' + self.version, 'PYTORCH_BUILD_NUMBER=1']
```

Comment:
Should we rename `options` to `default_opts`, to highlight that you can override them via the `custom_opts` easyconfig parameter?

Also add `VERBOSE=1`, perhaps only when `--debug` is used?

Reply:
Adding a comment above: `Gather default options. Will be checked against (and can be overwritten by) custom_opts`. IMO it isn't worth highlighting that those are default_opts as it reduces readability.

Where is "VERBOSE=1" coming from? Is this supported by PyTorch buildsystem?

Reply:
If it doesn't seem to do anything, ignore it, it may be an attempt at debugging to was left in place.

It was included in the PyTorch 1.4 easyconfig, see https://github.com/easybuilders/easybuild-easyconfigs/pull/11000/files#diff-5e116091061fb8243f4e109886113febL273

Reply:
I can try and see if it does anything

Reply:
Using `VERBOSE=1` results in compiler commands actually being shown, I checked that.

So adding `VERBOSE=1` when `--debug` is used via a3a8d7130